### PR TITLE
Add datetime values for datetime typed arguments

### PR DESCRIPTION
**Pull request by wachterjohannes**

### CODE REVIEW

File: src/Application/Writer/WriteVisitor.php [hunk 1]
```diff
@@ -21,6 +21,8 @@
  */
 final class WriteVisitor extends NodeVisitorAbstract
 {
+    private array $additionalUsages = [];
```
Comment (alexander-schranz) [inline src/Application/Writer/WriteVisitor.php]: why is this required. the writer shold always write FQCN like rector and php-cs-fixer / rector hook should convert that to `use` statements.

File: src/Application/Grouper/MethodGrouper.php [hunk 2]
```diff
@@ -51,7 +51,7 @@ public function groupMethods(array $methods): array
 
                 if (isset($methods[$setMethod])) {
                     $type = 'get_set';
-                    $name = 'testSetGet' . \substr($method, 3);
+                    $name = 'test' . \substr($method, 3);
```
Comment (alexander-schranz) [inline src/Application/Grouper/MethodGrouper.php]: this is currently testing `SetGet` thats why it is named this way. It should be as explicit as possible as there could be other methods for that attribute.

File: src/Application/Grouper/MethodGrouper.php [hunk 3]
```diff
@@ -65,7 +65,7 @@ public function groupMethods(array $methods): array
                 $getMethod = \str_replace('set', 'get', $method);
                 if (isset($methods[$getMethod])) {
                     $type = 'get_set';
-                    $name = 'testSetGet' . \substr($method, 3);
+                    $name = 'test' . \substr($method, 3);
```
Comment (alexander-schranz) [inline src/Application/Grouper/MethodGrouper.php]: same here this is testing currently `SetGet` and that is why it is named this way. It should be as explicit as possible as there could be other methods for that attribute.

File: src/Application/Generator/ArgumentGenerator.php [hunk 4]
```diff
@@ -37,7 +43,11 @@ public function generateArguments(array $methodAttributes, string $behaviour = '
 
                 continue;
             } elseif ($attributeConfig instanceof NullableType) {
-                $typeName = $attributeConfig->type->name;
+                if ($attributeConfig->type instanceof FullyQualified) {
+                    $typeName = $attributeConfig->type->toString();
```
Comment (alexander-schranz) [inline src/Application/Generator/ArgumentGenerator.php]: is this happening in the datetime case?

**Review by alexander-schranz — CHANGES_REQUESTED** [review]

File: src/Application/Generator/ArgumentGenerator.php [hunk 5]
```diff
@@ -60,6 +70,30 @@ public function generateArguments(array $methodAttributes, string $behaviour = '
                 }
 
                 $attribute = $floatAttributes[$attributeName];
+            } elseif ('DateTimeImmutable' === $typeName) {
```
Comment (alexander-schranz) [inline src/Application/Generator/ArgumentGenerator.php:76]: Test Case missing for DateTime and DateTimeImmutable

File: src/Application/Writer/WriteVisitor.php [hunk 6]
```diff
@@ -191,4 +193,9 @@ public function leaveNode(Node $node)
 
         return $node;
     }
+
+    public function getAdditionalUsages(): array
```
Comment (alexander-schranz) [inline src/Application/Writer/WriteVisitor.php]: this should be removed as this should be handled by php-cs-fixer/rector to convert FCQN to use statements.

File: src/Application/Generator/ArgumentGenerator.php (hunk 4 above)
Comment (wachterjohannes) [reply src/Application/Generator/ArgumentGenerator.php]: yes 👍

File: src/Application/Generator/ArgumentGenerator.php (hunk 5 above)
Comment (wachterjohannes) [reply src/Application/Generator/ArgumentGenerator.php:76]: done

File: src/Application/Writer/WriteVisitor.php (hunk 1 above)
Comment (wachterjohannes) [reply src/Application/Writer/WriteVisitor.php]: the writer does not write the FQCN to the output at the end. this is. the reason why i added the handling of the additional usage.

File: src/Application/Writer/WriteVisitor.php (hunk 6 above)
Comment (wachterjohannes) [reply src/Application/Writer/WriteVisitor.php]: that does not work with the writer of the library

File: src/Application/Generator/DateTimeGenerator.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Schranz\TestGenerator\Application\Generator;
+
+/**
+ * @internal
+ */
+class DateTimeGenerator
```
Comment (wachterjohannes) [inline src/Application/Generator/DateTimeGenerator.php]: needed to have fixed datetimes for the testcases

Comment (alexander-schranz) [reply src/Application/Generator/DateTimeGenerator.php]: it should not create 100% occur code would jsut create a datetime with `1970-01-01` always

Comment (alexander-schranz) [reply src/Application/Generator/DateTimeGenerator.php]: second datetime with `1970-01-02` like its done for the integers and floats

**Comment by wachterjohannes** [comment]
@alexander-schranz fixed

**Comment by wachterjohannes** [comment]
@alexander-schranz but there is a problem 🙈  The assertSame with a new Datetime does not work. See the following snippets what is generated currently and what it should be:

Generated but does not work:
```
    public function testSetGetBirthday(): void
    {
        $model = $this->createInstance();
        $this->assertNull($model->getBirthday());
        $this->assertSame($model, $model->setBirthday(new DateTimeImmutable('2022-01-01')));
        $this->assertSame(new DateTimeImmutable('2022-01-01'), $model->getBirthday());
        $this->assertSame($model, $model->setBirthday(null));
        $this->assertNull($model->getBirthday());
        $this->markAsRisky();
    }
```

What it should be
```
    public function testSetGetBirthday(): void
    {
        $model = $this->createInstance();
        $this->assertNull($model->getBirthday());
        $value = new DateTimeImmutable('2022-01-01');
        $this->assertSame($model, $model->setBirthday($value));
        $this->assertSame($value, $model->getBirthday());
        $this->assertSame($model, $model->setBirthday(null));
        $this->assertNull($model->getBirthday());
        $this->markAsRisky();
    }
```

**Comment by alexander-schranz** [comment]
@wachterjohannes that is already on my list the argument generator still should behave the same. At current state manual changes required here to make the test run.

File: src/Application/Writer/WriteVisitor.php (hunk 6 above)
Comment (alexander-schranz) [reply src/Application/Writer/WriteVisitor.php]: fixed this via rector. the writer need to use the `\` prefixed class else it tries to use a relative class to the current namespace.

**Review by alexander-schranz — APPROVED** [review]
